opta/linker.py

```diff
@@ -0,0 +1,416 @@
+from collections.abc import MutableSequence
+from dataclasses import dataclass, field
+from typing import Any, Dict, FrozenSet, Iterable, List, Set
+
+from opta.layer2 import Layer
+from opta.link import Link
+from opta.link_spec import InputLinkSpec, OutputLinkSpec
+from opta.module2 import Module
+from opta.module_spec import ModuleSpec
+from opta.utils.ref import (
+    InterpolatedReference,
+    Reference,
+    SimpleInterpolatedReference,
+    get_all_references,
+    parse_ref_string,
+)
+from opta.utils.visit import Visitor, fill_missing_list_or_dict
+
+_ModuleMap = Dict[str, Module]
+
+
+@dataclass
+class LinkResult:
+    execution_order: List[FrozenSet[str]] = field(default_factory=list)
+    interpolations: Dict[str, Dict[Reference, InterpolatedReference]] = field(
+        default_factory=dict
+    )
+
+
+class Linker:
```

Comment:
can this be covered by unit test?
sounds like a critical component of the new module api